tmol/score/omega/potentials/omega_pose_score.impl.hh

```diff
@@ -0,0 +1,176 @@
+
+template <typename Real, int N>
+using Vec = Eigen::Matrix<Real, N, 1>;
+template <typename Real>
+using CoordQuad = Eigen::Matrix<Real, 4, 3>;
+
+template <typename Real, typename Int, tmol::Device D>
+TMOL_DEVICE_FUNC int resolve_atom_from_uaid(
+    TensorAccessor<Int, 1, D> uaid,
+    int block_index,
+    int pose_index,
+
+    TView<Vec<Real, 3>, 2, D> coords,
+    TView<Int, 2, D> pose_stack_block_coord_offset,
+    TView<Int, 2, D> pose_stack_block_type,
+    TView<Vec<Int, 2>, 3, D> pose_stack_inter_block_connections,
+    TView<Int, 3, D> block_type_omega_quad_uaids,
+    TView<Int, 3, D> block_type_atom_downstream_of_conn) {
+  if (uaid[0] != -1) {  // This uaid resides in this block
+    int block_coord_offset =
+        pose_stack_block_coord_offset[pose_index][block_index];
+    return block_coord_offset + uaid[0];  // The actual global atom index
+  } else {                                // We need to follow to another block
+    int connection_index = uaid[1];
+    int sep = uaid[2];
+
+    const Vec<Int, 2>& connection =
+        pose_stack_inter_block_connections[pose_index][block_index]
+                                          [connection_index];
+    int other_block_index = connection[0];
+
+    if (other_block_index == -1) {
+      // This residue doesn't exist!
+      return -1;
+    }
+
+    int other_connection_index = connection[1];
+    int other_block_type_index =
+        pose_stack_block_type[pose_index][other_block_index];
+
+    int idx = block_type_atom_downstream_of_conn[other_block_type_index]
+                                                [other_connection_index]
+                                                [sep];  // The offset within the
+                                                        // other block
+    int block_coord_offset =
+        pose_stack_block_coord_offset[pose_index][other_block_index];
+    return block_coord_offset + idx;
+  }
+}
+
+template <
+    template <tmol::Device>
+    class DeviceDispatch,
+    tmol::Device D,
+    typename Real,
+    typename Int>
+auto OmegaPoseScoreDispatch<DeviceDispatch, D, Real, Int>::f(
+    TView<Vec<Real, 3>, 2, D> coords,
+    TView<Int, 2, D> pose_stack_block_coord_offset,
+    TView<Int, 2, D> pose_stack_block_type,
+    TView<Vec<Int, 2>, 3, D> pose_stack_inter_block_connections,
+    TView<Int, 3, D> block_type_omega_quad_uaids,
+    TView<Int, 3, D> block_type_atom_downstream_of_conn,
+
+    TView<OmegaGlobalParams<Real>, 1, D> global_params,
+
+    bool compute_derivs
+
+    ) -> std::tuple<TPack<Real, 2, D>, TPack<Vec<Real, 3>, 3, D>> {
+  int const n_poses = coords.size(0);
+  int const max_n_atoms = coords.size(1);
+  auto V_t = TPack<Real, 2, D>::zeros({1, n_poses});
+  auto dV_dx_t = TPack<Vec<Real, 3>, 3, D>::zeros({1, n_poses, max_n_atoms});
+
+  auto V = V_t.view;
+  auto dV_dx = dV_dx_t.view;
+
+  // Optimal launch box on v100 and a100 is nt=32, vt=1
+  LAUNCH_BOX_32;
+  // Define nt and reduce_t
+  CTA_REAL_REDUCE_T_TYPEDEF;
```

Comment:
What do these do? Are these appropriate for omega?

Reply:
Is the 32-threads-per-CTA going to be spmething that is energy term specific? Do we want this centrally defined so that we can update it more easily depending on the backend hardware?